cpp/serve/sampler.cc

```diff
@@ -375,62 +375,66 @@ class CPUSampler : public SamplerObj {
-            }
-
-            // sample a new token from the new distribution
-            int32_t new_token =
-                SampleTopPFromProb(logits_or_probs_on_cpu, verify_start + cur_token_idx,
-                                   generation_cfg[i]->top_p, rngs[i]->GetRandomNumber())
-                    .second;
-            request_mstates[i]->CommitToken(new_token);
-            accepted_tokens[i].push_back(cur_token);
-            break;
-          }
-        },
-        0, num_sequence);
+    auto f_task = [&](int i) {
+      int verify_start = cum_verify_lengths[i];
+      int verify_end = cum_verify_lengths[i + 1];
+      for (int cur_token_idx = 0; cur_token_idx < verify_end - verify_start; ++cur_token_idx) {
+        if (!can_compute_prob_in_parallel) {
+          SinglePosComputeProbsFromLogitsInplace(logits_or_probs_on_cpu,
+                                                 verify_start + cur_token_idx, request_mstates[i],
+                                                 generation_cfg[i]);
+        }
+
+        float* p_probs = global_p_probs + (verify_start + cur_token_idx) * vocab_size;
+        int cur_token = draft_output_tokens[i][cur_token_idx];
+        float q_value = draft_output_token_prob[i][cur_token_idx];
+        float p_value = p_probs[cur_token];
+
+        if (p_value >= q_value) {
+          request_mstates[i]->CommitToken(cur_token);
+          accepted_tokens[i].push_back(cur_token);
+          continue;
+        }
+        float r = rngs[i]->GetRandomNumber();
+        if (r < p_value / (q_value + eps_)) {
+          request_mstates[i]->CommitToken(cur_token);
+          accepted_tokens[i].push_back(cur_token);
+          continue;
+        }
+
+        // normalize a new probability distribution
+        double sum_v = 0.0;
+        NDArray q_dist = draft_output_prob_dist[i][cur_token_idx];
+        ICHECK(q_dist->device.device_type == kDLCPU);
+        ICHECK(q_dist->ndim == 1);
+        ICHECK(vocab_size == q_dist->shape[q_dist->ndim - 1]);
+        const float* __restrict p_qdist =
+            static_cast<float*>(__builtin_assume_aligned(q_dist->data, 4));
+
+        for (int j = 0; j < vocab_size; ++j) {
+          p_probs[j] = std::max(p_probs[j] - p_qdist[j], 0.0f);
+          sum_v += p_probs[j];
+        }
+        for (int j = 0; j < vocab_size; ++j) {
+          p_probs[j] /= sum_v;
+        }
+
+        // sample a new token from the new distribution
+        int32_t new_token = SampleTopPFromProb(logits_or_probs_on_cpu, verify_start + cur_token_idx,
+                                               generation_cfg[i]->top_p, rngs[i]->GetRandomNumber())
+                                .second;
+        request_mstates[i]->CommitToken(new_token);
+        accepted_tokens[i].push_back(cur_token);
+        break;
+      }
+    };
+
+    if (num_sequence == 1) {
+      f_task(0);
+    } else {
+      tvm::runtime::parallel_for_with_threading_backend(std::move(f_task), 0, num_sequence);
```

Comment:
@tqchen May I have a quick question here? I am wondering if the template expansion trick still works when using `std::move` here.